### PR TITLE
Feature/ex 3547 implementation

### PR DESCRIPTION
**Pull request by davidmfempathy**
# Description

USER STORY: https://searchbroker.atlassian.net/browse/EX-3546

The description of the user story is a little bit outdated. This task could be split in different smaller tasks:

- Modify util currency-formatter: remove `hideIntegerDecimals` parameter. The regex has been modified and now the format accepts a `?` symbol to remove the non-significant zeroes if they are not needed. E.g: number: 1234 -> i.iii,?dd = 1.234; i.iii,dd = 1.234,00.
- Add 'currency' to the 'snippetConfig' of X-Components which accepts a string with the current currency. E.g. 'EUR', 'USD', etc.
- Modify `BaseCurrency` to allow injection for the `currencyFormat`. Now the component can use injection or the prop.
- Remove required and default property 'format' for components that use `BaseCurrency` nested.

We'll need to release this in order to have it available in the X-Archetype which will have another PR.

X-Archetype will have a local configuration for the currencies:

```
const currencies: Record<string, string> = {
  EUR: 'i.iii,dd €',
  USD: '$i,iii.dd'
};
```

And then, the X-Archetype injects, using @XInject with the name "currentFormat", from the App to the rest of nested components, the format of the current currency.

This will be also available in the reactive config (snippetConfig).

# Test

- Check all the **TESTS** and **DOCUMENTATION** of the modified components.
- Check the functionality and tests of the new features.
- You can also build X-Components and import it in the X-Archetype. This feature is WIP in another branch from the X-Archetype but you can easily add this to App.vue component:

```
    const currencies: Record<string, string> = {
      EUR: 'i.iii,dd €',
      USD: '$i,iii.dd'
    };

    @XProvide('currencyFormat')
    public get currencyFormat(): string {
      return currencies[this.snippetConfig.currency!];
    }
```

Then you can:
- Modify the format of the current, which is 'EUR' by default.
- Modify it from the console to see the reactive config in action: `window.initX.currency = "USD"`.

NOTE: this approach is temporal. There'd be a intermediate project between X-Components and the X-Archetype that will have these default configuration.

**THIS IS A BREAKING CHANGE**

### CODE REVIEW

File: packages/x-components/src/views/base-config.ts [hunk 1]
```diff
@@ -19,5 +19,6 @@ export const baseInstallXOptions: InstallXOptions = {
         identifierDetectionRegexp: '^[a-zA-Z][0-9]+'
       }
     }
-  }
+  },
+  app: {}
```
Comment (pmareke) [inline packages/x-components/src/views/base-config.ts:23]: Why do we need this empty `app` property?

File: packages/x-components/src/components/currency/__tests__/base-currency.spec.ts [hunk 2]
```diff
@@ -126,15 +113,46 @@ describe('testing BaseCurrency component', () => {
...
+          };
+        }
+      }
+    );
+
+    expect(wrapper.text()).toBe('$12,345,678.876');
+  });
 });
 
 interface RenderBaseCurrencyOptions {
```
Comment (pmareke) [inline packages/x-components/src/components/currency/__tests__/base-currency.spec.ts:155]: Please document the interface and the props, please.

File: packages/x-components/src/components/currency/base-currency.vue [hunk 3]
```diff
@@ -6,11 +6,28 @@
...
   import { Component, Prop } from 'vue-property-decorator';
   import { currencyFormatter } from '../../utils/currency-formatter';
+  import { XInject } from '../decorators/injection.decorators';
 
   /**
-   * Renders the value received as a property, which always must be a JavaScript number, with the
-   * proper format provided as string property. The rendered tag is a span in order to render a
-   * default inline HTML element.
+   * Renders the value received as a property which always must be a JavaScript number, with the
+   * proper format provided as string property or by injection. The rendered tag is a span in order
```
Comment (pmareke) [inline packages/x-components/src/components/currency/base-currency.vue:13]: `* proper format provided as a string property or by injection. The rendered tag is a span in order`

File: packages/x-components/src/components/currency/base-currency.vue [hunk 4]
```diff
@@ -6,11 +6,28 @@
...
+   * to render a default inline HTML element.
+   *
+   * Regarding the format or mask to be defined as string:
+   * - Use 'i' to define integer numbers.
+   * - Use 'd' to define decimal numbers. You can define the length of the decimal part. If the
+   * doesn't include decimals, it is filled with zeros until reach the length defined with 'd's.
```
Comment (pmareke) [inline packages/x-components/src/components/currency/base-currency.vue:19]: ` * format doesn't include decimals, it is filled with zeros until reach the length defined with 'd's.`

File: packages/x-components/src/components/currency/base-currency.vue [hunk 5]
```diff
@@ -6,11 +6,28 @@
...
+   * - Integer separator must be defined between the 3rd and the 4th integer 'i' of a group.
+   * - Decimal separator must be defined between the last 'i' and the first 'd'. It can be more
+   * than one character.
+   * - If you want to hide the decimal part if it's zero, you can add the `?` symbol after the
+   * decimal separator (e.g. 'i.iii,?dd', for `1234` you would get `1.234` instead of `1.234,00`).
+   * - Set whatever you need around the integers and decimals marks.
+   * - Default mask: 'i.iii,dd' which returns '1.345,67'.
+   *
+   * @remarks The number of 'd', which is the maximum decimal length, MUST match with the length
```
Comment (pmareke) [inline packages/x-components/src/components/currency/base-currency.vue:28]: `MUST matches with the length`

File: packages/x-components/src/components/currency/base-currency.vue [hunk 6]
```diff
@@ -6,11 +6,28 @@
...
+   * of decimals provided from the adapter. Otherwise, when the component truncate the decimal
+   * part, delete significant digits.
```
Comment (pmareke) [inline packages/x-components/src/components/currency/base-currency.vue:30]: `* part, it will delete significant digits.`

**Review by pmareke — CHANGES_REQUESTED** [review]

File: packages/x-components/src/views/base-config.ts (hunk 1 above)
Comment (davidmfempathy) [reply packages/x-components/src/views/base-config.ts:23]: I missed it.

File: packages/x-components/src/components/currency/base-currency.vue (hunk 6 above)
Comment (davidmfempathy) [reply packages/x-components/src/components/currency/base-currency.vue:30]: I put it in present simple.

File: packages/x-components/src/utils/currency-formatter.ts [hunk 7]
```diff
@@ -55,21 +56,19 @@ interface NumberParts {
...
 
   const formattedInteger = formatInteger(integer, integerSeparator);
   const formattedDecimalsSeparator = formatDecimalSeparator(
-    decimalsNumber,
+    decimal,
     decimalSeparator,
     hideIntegerDecimals
   );
   const formattedDecimal = formatDecimal(decimal, decimalsNumber, hideIntegerDecimals);
+  debugger;
```
[the remark this answers is not in the document]
Comment (davidmfempathy) [reply packages/x-components/src/utils/currency-formatter.ts:71]: Sorry 🔫

File: packages/x-components/src/utils/__tests__/currency-formatter.spec.ts [hunk 8]
```diff
@@ -25,8 +25,14 @@ describe('testing Currency formatter', () => {
...
 
-  it('allows to hide decimals', () => {
-    expect(format(1234.56, 'i.iii,d', true)).toBe('1.234');
+  it('allows to hide the decimal part', () => {
+    expect(format(1234.56, 'i.iii')).toBe('1.234');
+  });
+
+  it("allows to hide the decimal part if it's zero with the '?' symbol within the format", () => {
+    expect(format(1234, 'i.iii,dd')).toBe('1.234,00');
+    expect(format(1234, 'i.iii,?dd')).toBe('1.234');
```
[the remark this answers is not in the document]
Comment (davidmfempathy) [reply packages/x-components/src/utils/__tests__/currency-formatter.spec.ts:34]: Yes, we also talked about it during the development but I forgot to move it.

File: packages/x-components/src/utils/__tests__/currency-formatter.spec.ts [hunk 9]
```diff
@@ -47,7 +53,7 @@ describe('testing Currency formatter', () => {
     expect(format(123456)).toBe('123.456,00');
   });
 
-  function format(value: number, format?: string, hideIntegerDecimals?: boolean): string {
-    return currencyFormatter(value, format, hideIntegerDecimals);
+  function format(value: number, format?: string): string {
```
[the remark this answers is not in the document]
Comment (davidmfempathy) [reply packages/x-components/src/utils/__tests__/currency-formatter.spec.ts:56]: Agree, it was done. I'm removing it.

**Review by javieri-empathy — APPROVED** [review]

**Review by pmareke — APPROVED** [review]

**Pull request merged**